forc-plugins/forc-crypto/src/keys/vanity.rs

```diff
@@ -0,0 +1,440 @@
+}
+
+pub trait VanityMatcher: Send + Sync {
+    fn is_match(&self, addr: &Address) -> bool;
+}
+
+struct HexMatcher {
+    left: Vec<u8>,
+    right: Vec<u8>,
+}
+
+impl VanityMatcher for HexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        bytes.starts_with(&self.left)
+            && bytes[bytes.len() - self.right.len()..]
+                .iter()
+                .zip(self.right.iter())
+                .all(|(a, b)| a.eq_ignore_ascii_case(b))
+    }
+}
+
+struct LeftHexMatcher {
+    left: Vec<u8>,
+}
+
+impl VanityMatcher for LeftHexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        bytes.starts_with(&self.left)
+    }
+}
+
+struct RightHexMatcher {
+    right: Vec<u8>,
+}
+
+impl VanityMatcher for RightHexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        bytes[bytes.len() - self.right.len()..]
+            .iter()
+            .zip(self.right.iter())
+            .all(|(a, b)| a.eq_ignore_ascii_case(b))
+    }
+}
+
+struct LeftExactRightRegexMatcher {
+    left: Vec<u8>,
+    right: Regex,
+}
+
+impl VanityMatcher for LeftExactRightRegexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        bytes.starts_with(&self.left) && self.right.is_match(&hex::encode(bytes))
+    }
+}
+
+struct LeftRegexRightExactMatcher {
+    left: Regex,
+    right: Vec<u8>,
+}
+
+impl VanityMatcher for LeftRegexRightExactMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        self.left.is_match(&hex::encode(bytes))
+            && bytes[bytes.len() - self.right.len()..]
+                .iter()
+                .zip(self.right.iter())
+                .all(|(a, b)| a.eq_ignore_ascii_case(b))
+    }
+}
+
+struct SingleRegexMatcher {
+    re: Regex,
+}
+
+impl VanityMatcher for SingleRegexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let addr = hex::encode(addr.as_ref());
+        self.re.is_match(&addr)
+    }
+}
+
+struct RegexMatcher {
+    left: Regex,
+    right: Regex,
+}
+
+impl VanityMatcher for RegexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let addr = hex::encode(addr.as_ref());
+        self.left.is_match(&addr) && self.right.is_match(&addr)
+    }
+}
+
+fn parse_pattern(pattern: &str, is_start: bool) -> Result<Either<Vec<u8>, Regex>> {
+    if let Ok(decoded) = hex::decode(pattern) {
+        if decoded.len() > 32 {
+            return Err(anyhow!("Hex pattern must be less than 32 bytes"));
+        }
+        Ok(Either::Left(decoded))
+    } else {
+        // Check if the pattern contains only valid hex characters
+        if !pattern
+            .chars()
+            .all(|c| c.is_ascii_hexdigit() || c == '^' || c == '$')
+        {
+            return Err(anyhow!("Invalid hex pattern: {}", pattern));
+        }
+
```

Comment:
From here we know it's a regex. Is there any need to touch it futher? i.e if the user is dropping down to a regex should we just let them match for whatever they're looking for and from here just pass it through?

Reply:
addressed: https://github.com/FuelLabs/sway/pull/6661/commits/59ab719069be67c3a2ef209e632c2d5039996746